cola/widgets/toolbar.py

```diff
@@ -0,0 +1,527 @@
+    ACTION_GREP: N_('Grep'),
+    ACTION_SEARCH: N_('Search...'),
+
+    ACTION_STAGE: N_('Stage'),
+    ACTION_AMMEND_LAST: N_('Amend Last Commit'),
+    ACTION_STAGE_ALL: N_('Stage All Untracked'),
+    ACTION_UNSTAGE_ALL: N_('Unstage All'),
+    ACTION_UNSTAGE: N_('Unstage'),
+    ACTION_LOAD_COMMIT: N_('Load Commit Message...'),
+    ACTION_APPLY_COMMIT: N_('Get Commit Message Template'),
+
+    ACTION_EXPRESSION: N_('Expression...'),
+    ACTION_BRANCHES: N_('Branches...'),
+    ACTION_DIFFSTAT: N_('Diffstat'),
+
+    ACTION_REVIEW: N_('Review...'),
+    ACTION_CREATE: N_('Create...'),
+    ACTION_CHECKOUT: N_('Checkout...'),
+    ACTION_DELETE: N_('Delete...'),
+    ACTION_DELETE_REMOTE: N_('Delete Remote Branch...'),
+    ACTION_RENAME: N_('Rename Branch...'),
+    ACTION_BROWSE_CURRENT: N_('Browse Current Branch...'),
+    ACTION_BROWSE_OTHER: N_('Browse Other Branch...'),
+    ACTION_VISUALIZE_CURRENT: N_('Visualize Current Branch...'),
+    ACTION_VISUALIZE_ALL: N_('Visualize All Branches...'),
+
+    ACTION_REBASE: N_('Start Interactive Rebase...'),
+    ACTION_REBASE_EDIT: N_('Edit...'),
+    ACTION_REBASE_CONT: N_('Continue'),
+    ACTION_REBASE_IGNORE: N_('Skip Current Patch'),
+    ACTION_REBASE_ABORT: N_('Abort')
+}
+TOOLBAR_ICONS = {
+    ACTION_NEW_REPO: icons.new(),
+    ACTION_OPEN_REPO: icons.folder(),
+    ACTION_OPEN_NEW_WIN: icons.folder(),
+    ACTION_CLONE: icons.repo(),
+    ACTION_REFRESH: icons.sync(),
+    ACTION_FIND_FILES: icons.zoom_in(),
+    ACTION_SAVE_TAR_ZIP: icons.file_zip(),
+    ACTION_PREFERENCES: icons.configure(),
+
+    ACTION_PULL: icons.pull(),
+    ACTION_PUSH: icons.push(),
+    ACTION_CREATE_TAG: icons.tag(),
+    ACTION_MERGE: icons.merge(),
+    ACTION_SEARCH: icons.search(),
+
+    # TODO: add more icons
+}
+
+
+def create_toolbar(toolbar):
+    toolbar.setToolButtonStyle(Qt.ToolButtonIconOnly)
+    toolbar.clear()
+
+    actions_dict = load_toolbar_config()
+    for action_dict in actions_dict:
+        add_toolbar_action(toolbar, action_dict)
+
+
+def add_toolbar_action(toolbar, action_dict):
+    if action_dict['title'] == SEPARATOR_TEXT:
+        toolbar.addSeparator()
+    else:
+        callback = TOOLBAR_ACTIONS[action_dict['parent']][action_dict['action']]
+        if action_dict['icon'] in TOOLBAR_ICONS:
+            icon = TOOLBAR_ICONS[action_dict['icon']]
+            toolbar_action = toolbar.addAction(icon, action_dict['title'],
+                                               callback)
+        else:
+            toolbar_action = toolbar.addAction(action_dict['title'], callback)
+
+        toolbar_action.setData(action_dict)
+
+
+def load_toolbar_config():
+    prefs_model = prefs.PreferencesModel()
+    return json.loads(prefs_model.get_config(SOURCE_CONFIG, PREFS_TOOLBAR))
+
+
+def save_toolbar_config(actions_dict):
+    prefs_model = prefs.PreferencesModel()
+    cmds.do(prefs.SetConfig, prefs_model, SOURCE_CONFIG, PREFS_TOOLBAR,
+            json.dumps(actions_dict))
```

Comment:
Loading and saving the toolbar config here is slow because we have to touch the filesystem (again).  We shouldn't do that because we already have code paths that save and restore "settings", and we should leverage that.

Also, this assumes that we only have a single toolbar, which is not necessarily true.  `git-dag` might have its own.  It's better to tie the save/restore to the main widget's save/restore, and have it push values to the toolbar instances that it owns.

The functions in the main widget related to saving state are `export_state()` and `apply_state()`.